component/commons/handler/src/main/java/org/wso2/carbon/identity/password/expiry/validation/handler/PasswordExpiryValidationHandler.java

```diff
@@ -0,0 +1,146 @@
+ */
+public class PasswordExpiryValidationHandler extends AbstractEventHandler {
+    private static final Log log = LogFactory.getLog(PasswordExpiryValidationHandler.class);
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        String username = (String) event.getEventProperties().get(IdentityEventConstants.EventProperty.USER_NAME);
+        String tenantDomain = (String) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.TENANT_DOMAIN);
+        UserStoreManager userStoreManager = (UserStoreManager) event.getEventProperties()
+                .get(IdentityEventConstants.EventProperty.USER_STORE_MANAGER);
+        String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+
+        try {
+            if (isPasswordExpired(tenantDomain, tenantAwareUsername, userStoreManager)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User: " + username + " password is expired.");
+                }
+                throw new IdentityEventException("Password expired for user with username: " + tenantAwareUsername);
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityEventException("UserStore Exception occurred", e);
+        } catch (ParseException e) {
+            throw new IdentityEventException("Parse Exception occurred", e);
+        }
+    }
+     /**
+     * Check whether the password is expired or not.
+     * @param tenantDomain
+     * @param tenantAwareUsername
+     * @param userStoreManager
+     * @return true if password is expired or password last update time is null
+     * @throws UserStoreException
+     * @throws ParseException
+     */
+    private boolean isPasswordExpired(String tenantDomain, String tenantAwareUsername,
+                                      UserStoreManager userStoreManager)
+            throws UserStoreException, ParseException {
+
+        String passwordLastChangedTime = getClaimValue(userStoreManager,
+                PasswordExpiryValidationConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM, tenantAwareUsername);
+
+        if (passwordLastChangedTime == null) {
+            String createdClaimValue = getClaimValue(userStoreManager,
+                    PasswordExpiryValidationConstants.CREATED_CLAIM, tenantAwareUsername);
+            if (createdClaimValue != null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("User: " + tenantAwareUsername + " Claim: "
+                            + PasswordExpiryValidationConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM +
+                            " is null.Hence considered claim:" + PasswordExpiryValidationConstants.CREATED_CLAIM);
+                }
+                passwordLastChangedTime = convertCreatedDateToEpochString(createdClaimValue);
+            }
```

Comment:
If the user do not have a password changed time we are to prompt the user for password update. We can use the same logic used in here
https://github.com/wso2-extensions/identity-outbound-auth-passwordPolicy/blob/abbf0760f31aa02c7522f6bfd121945c6c611bdb/component/authenticator/src/main/java/org/wso2/carbon/identity/policy/password/PasswordResetEnforcer.java#L264

Reply:
The as the time created claim is a user claim it could introduce complexities to the flow.

Reply:
@IsuraD wdyt?

Reply:
Used http://wso2.org/claims/lastPasswordChangedTimestamp to have backward compatibility as per the issue [1].  Removed the created time claim which was used to check the last password update time as per the design doc.

[1] - https://github.com/wso2/product-is/issues/12690